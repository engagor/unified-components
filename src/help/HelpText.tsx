import React, {PropsWithChildren, ReactElement} from 'react';
import { Options } from 'tooltip.js';
import Tooltip from "../utils/Tooltip";

export type HelpTextProps = PropsWithChildren<{
    text: string;
    color?: string;
    id?: string;
    placement?: Options['placement'];
    className?: string;
}>;

const defaultProps: HelpTextProps = {
    text: '',
    color: 'black',
    id: '',
    placement: 'right',
};

export default function HelpText(props: HelpTextProps): ReactElement {
    const getTemplate = (): string => {
        return `
            <div class="popover bg-white box-shadow-default rounded-4 w-max-280 p-8" role="tooltip">
                <div class="tooltip-arrow arrow"></div>
                <div class="popover-content">
                    <div class='tooltip-inner'></div>
                </div>
            </div>
        `;
    };

    // We do this extra guard because if someone sends null or empty string,
    // React's default props will not kick in (only on undefined).
    let placementString = props.placement;
    if (!placementString) {
        placementString = defaultProps.placement;
    }

    const classNames = [
        'd-flex',
        'align-items-center',
        'justify-center',
        'bg-charcoal-color-200',
        'rounded-full',
        'text-white',
        'font-bold',
        'w-16',
        'h-16',
        'text-sm-2',
        'cursor-pointer',
    ];

    let content = (
        <span className={`${classNames.join(' ')} inline_help_${props.color} ${props.className || ''}`} id={`help_inline_${props.id}`}>
            ?
        </span>
    );

    if (props.children) {
        content = <span className="tooltip-wrapper">{props.children}</span>;
    }

    return (
        <Tooltip
            container={document.body}
            title={props.text}
            html={true}
            template={getTemplate()}
            placement={placementString}
        >
            {content}
        </Tooltip>
    );
};

HelpText.defaultProps = defaultProps;
