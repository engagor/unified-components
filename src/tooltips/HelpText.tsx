import React, { isValidElement, PropsWithChildren, ReactElement, useState } from 'react';
import Tippy from '@tippyjs/react/headless';
import Arrow from './Arrow';
import Placement from './Placement';

export type HelpTextProps = PropsWithChildren<{
    title?: string | ReactElement;
    content: string | ReactElement;
    appendTo?: 'parent' | Element | ((ref: Element) => Element)
    placement?: Placement;
    visible?: boolean;
}>;

export default function HelpText({ title, content, appendTo, placement, visible, children }: HelpTextProps): ReactElement | null {
    const [arrow, setArrow] = useState<HTMLDivElement | null>(null);

    if (!appendTo) {
        appendTo = document.body;
    }

    if (!placement) {
        placement = 'right';
    }

    if (!children) {
        const classNames = [
            'text-charcoal-color-200',
            'w-16',
            'h-16',
            'cursor-pointer',
        ];

        children = (
            <span className={`${classNames.join(' ')}`}>
                <span aria-hidden={true} className="en-icon en-icon-help"></span>
            </span>
        );
    }

    if (!isValidElement(children)) {
        return null;
    }

    const popperOptions = {
        modifiers: [
            { name: 'arrow', options: { element: arrow } },
        ],
    };

    return (
        <Tippy
            placement={placement}
            visible={visible}
            appendTo={appendTo}
            popperOptions={popperOptions}
            render={
                (attrs) => {
                    if (title) {
                        title = <h2 className="text-base font-bold m-0 mb-8">{title}</h2>;
                    }

                    return (
                        <article className="shadow bg-white dark:bg-charcoal-color-400 dark:text-white dark:border-color-400 rounded-4 max-w-280 p-16" role="tooltip" {...attrs}>
                            {title}
                            <section className="text-base font-normal">{content}</section>
                            <Arrow className="text-white dark:text-charcoal-color-400" ref={(node) => setArrow(node)} />
                        </article>
                    );
                }
            }
        >
            {children}
        </Tippy>
    );
}
