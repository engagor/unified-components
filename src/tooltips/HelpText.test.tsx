import { renderToStaticMarkup } from 'react-dom/server';

jest.mock(
    'react-dom',
    () => {
        return {
            createPortal: (element: ReactElement, node: HTMLElement) => {
                node.innerHTML = renderToStaticMarkup(element);
            },
        }
    }
);

import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import HelpText from './HelpText';

it('renders correctly', async () => {
    const element = document.createElement('div');

    const helptext = await renderer
        .create(<HelpText content="Hello this is my awesome helptext!" visible={true} appendTo={element} />)
        .toJSON();

    await expect(helptext).toMatchSnapshot();
    expect(element).toMatchSnapshot();
});
