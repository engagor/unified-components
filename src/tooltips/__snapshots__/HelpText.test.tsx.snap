// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span
  className="flex items-center justify-center bg-charcoal-color-200 rounded-full text-white font-bold w-16 h-16 text-sm-2 cursor-pointer"
>
  ?
</span>
`;

exports[`renders correctly 2`] = `
<div>
  <div
    data-tippy-root=""
    id="tippy-1"
    style="pointer-events: none; z-index: 9999; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(10px, 0px);"
  >
    <article
      class="bg-white box-shadow-default rounded-4 w-max-280 p-16"
      data-escaped=""
      data-placement="right"
      data-reference-hidden=""
      role="tooltip"
    >
      <section
        class="text-base font-normal"
      >
        Hello this is my awesome helptext!
      </section>
      <div
        class="arrow "
      />
    </article>
  </div>
</div>
`;
